### PR TITLE
Statevector vjp and support Hamiltonian in adjoint_jacobian

### LINKED ISSUE #216: Add support for `qml.Hamiltonian` with `lightning.qubit` for `diff_method='adjoint'`
#### Issue description

*Description of the issue:*
When running the following simple code block, it throws the `NotImplementedError`. There is no issue if I change to other simulators such as `default.qubit`, or replace the observable to simple observables such as `qml.PauliX(0)`.

```
import pennylane as qml
import numpy as np

dev = qml.device("lightning.qubit", wires=4)
H, qubits = qml.qchem.molecular_hamiltonian(['H','H'], np.array([0.0, 0.1, 0.0, 0.0, -0.1, 0.0]))

@qml.qnode(dev, diff_method='adjoint')
def circuit(params):
    qml.RX(params, wires=1)
    return qml.expval(H)

circuit(0.123)
```

* *Expected behavior:* It returns the expectation value of the Hamiltonian.

* *Actual behavior:*  It throws `NotImplementedError`.

* *Reproduces how often:* Every time

* *System information:*

Name: PennyLane
Version: 0.20.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /usr/local/lib/python3.7/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, scipy, semantic-version, toml
Required-by: amazon-braket-pennylane-plugin, PennyLane-Lightning, PennyLane-Qchem
Platform info:           Linux-4.14.256-197.484.amzn2.x86_64-x86_64-with-debian-buster-sid
Python version:          3.7.10
Numpy version:           1.19.2
Scipy version:           1.5.2
Installed devices:
- default.gaussian (PennyLane-0.20.0)
- default.mixed (PennyLane-0.20.0)
- default.qubit (PennyLane-0.20.0)
- default.qubit.autograd (PennyLane-0.20.0)
- default.qubit.jax (PennyLane-0.20.0)
- default.qubit.tf (PennyLane-0.20.0)
- default.qubit.torch (PennyLane-0.20.0)
- lightning.qubit (PennyLane-Lightning-0.20.2)
- braket.aws.qubit (amazon-braket-pennylane-plugin-1.3.0)
- braket.local.qubit (amazon-braket-pennylane-plugin-1.3.0)

### PR DESCRIPTION
**Pull request by chaeyeunpark**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

**Description of the Change:** We now have `pennylane_lightning.lightning_ops.adj_diff` module that handles all observables related to adjoint differentiation method. We also have adjoint-like computation of the statevecor vjp. 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:** Closes #216

### CODE REVIEW

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (mlxd) [inline pennylane_lightning/src/algorithms/Observables.hpp:132]: Do we want this to be final? Is there any advantage to it?

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (mlxd) [inline pennylane_lightning/src/algorithms/Observables.hpp:187]: Ownership of the tensor is maintained at the Python layer?
We are guaranteed the `shared_ptr` is a `shared_ptr` all the way through?

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (mlxd) [inline pennylane_lightning/src/algorithms/Observables.hpp:286]: I assume this will be run if available?

File: pennylane_lightning/src/algorithms/StateVecAdjDiff.hpp
Comment (mlxd) [inline pennylane_lightning/src/algorithms/StateVecAdjDiff.hpp:49]: This function looks great!

**Review by mlxd — COMMENTED** [review]
Awesome work @chaeyeunpark 
I have a few questions and comments:

**Comment by chaeyeunpark** [comment]
@mlxd Thank you for your prompt comments! I will resolve those issues soon as well as adding more tests for codecov.

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp:187]: We explicitly expose `std::shared_ptr` of observable classes with pybind11 (https://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html#std-shared-ptr).

File: pennylane_lightning/lightning_qubit.py
Comment (chaeyeunpark) [reply pennylane_lightning/lightning_qubit.py:189]: I think this is from the old attempts to reduce copy. As I think they are not really different, I will just revert this change.

File: pennylane_lightning/lightning_qubit.py
Comment (chaeyeunpark) [reply pennylane_lightning/lightning_qubit.py:323]: Let's just keep this.

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp:44]: Oops

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp:46]: Done

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp:169]: I think copy is fine here (as it has at most 30-40 elements).

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp:286]: I think so. Let me check.

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp:141]: I think this class (currently) strictly models a Hermitian matrix whose data resides on CPU memory. For integrating GPU, I think we may (1) add another abstraction of data (that can be either CPU or GPU data) or (2) implement `HermitianObsGPU` class (and also rename the current one to `HermitianObsCPU`).

Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp:141]: As there is also an issue regarding `applyInPlace` function, I think (1) is better in a long run. I will summarize my idea soon in Slack.

File: pennylane_lightning/lightning_qubit.py
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit.py:233]:
```suggestion
            Expectation or State: a common return type of measurements.
```

File: pennylane_lightning/lightning_qubit.py
Comment (AmintorDusko) [inline pennylane_lightning/lightning_qubit.py:392]:
```suggestion
            measurements (list): List of measurement processes for vector-Jacobian product
```

File: pennylane_lightning/src/algorithms/AdjointDiff.hpp
Comment (AmintorDusko) [inline pennylane_lightning/src/algorithms/AdjointDiff.hpp:120]:
```suggestion
```

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (AmintorDusko) [inline pennylane_lightning/src/algorithms/Observables.hpp:75]:
```suggestion
     * @brief Test whether this object is different from another object.
```

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (AmintorDusko) [inline pennylane_lightning/src/algorithms/Observables.hpp:68]:
```suggestion
     * @brief Test whether this object is equal to another object
```

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (AmintorDusko) [inline pennylane_lightning/src/algorithms/Observables.hpp:158]:
```suggestion
     * @param wires Wires the observable applies to.
```

File: pennylane_lightning/src/algorithms/AlgUtil.hpp
Comment (mlxd) [reply pennylane_lightning/src/algorithms/AlgUtil.hpp:136]: That was an old decision I made to stop the improper formatting of the openmp pragmas. There was a fix due in newer versions, but I believe we do not use those yet.

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (AmintorDusko) [inline pennylane_lightning/src/algorithms/Observables.hpp:161]: Could you, please, change `arg1` to a meaningful name?

File: pennylane_lightning/src/algorithms/AlgUtil.hpp
Comment (AmintorDusko) [reply pennylane_lightning/src/algorithms/AlgUtil.hpp:136]: Sure. Thanks for clarifying @mlxd.

**Comment by AmintorDusko** [comment]
Hi @chaeyeunpark. I'm suggesting some fixes mostly for typos. 
I reviewed close to half of the PR.
I'll be back to it after the PR is marked as ready for review again.
Nice work so far!

File: Makefile
Comment (chaeyeunpark) [reply Makefile:79]: This is because we now have a separate `test-cpp-omp`. Still, I think we may enable OpenMP here and make `test-cpp-no-omp`

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp:288]: Yes this version is used when OpenMP is enabled.

File: pennylane_lightning/src/algorithms/Observables.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/Observables.hpp]: Thanks for pointing this out! It is not updated.

File: pennylane_lightning/src/algorithms/AdjointDiff.hpp
Comment (AmintorDusko) [inline pennylane_lightning/src/algorithms/AdjointDiff.hpp:128]:
```suggestion
```

Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/AdjointDiff.hpp:128]: Hi Amintor, thanks for the comment. I found that this line has very weird property that `clang-format` (version 12 we are using) does not correctly format the line above (the line with `// clang-format off`) without this blank. It must from a bug in `clang-format`, but I think we should keep this line at the moment (until we upgrade clang-format version).

Comment (AmintorDusko) [reply pennylane_lightning/src/algorithms/AdjointDiff.hpp:128]: Oh! Weird! Thanks for the clarification!

File: tests/test_adjoint_jacobian.py
Comment (AmintorDusko) [inline tests/test_adjoint_jacobian.py:101]:
```suggestion
        """Tests if an empty array is returned when the measurements of the tape is empty."""
```

File: tests/test_adjoint_jacobian.py
Comment (AmintorDusko) [inline tests/test_adjoint_jacobian.py:859]: I think this sentence needs to be reviewed.

File: pennylane_lightning/src/tests/VectorStream.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/tests/VectorStream.hpp:1]: The original purpose I remember is that one in `Util.hpp` is inside `Pennylane::Util` namespace (instead of global) and I want to reduce dependencies in a test code. Still, as it seems like we do not use this file at the moment, I think we may remove it.

File: tests/test_vjp.py
Comment (AmintorDusko) [inline tests/test_vjp.py:84]: Maybe this sentence could use some review too.

**Comment by AmintorDusko** [comment]
Hi @chaeyeunpark. Once again, nice work so far!

File: tests/test_vjp.py
Comment (chaeyeunpark) [reply tests/test_vjp.py]: Yes, I also updated this one.

File: pennylane_lightning/src/algorithms/JacobianTape.hpp
Comment (mlxd) [inline pennylane_lightning/src/algorithms/JacobianTape.hpp:241]: I assume the `shared_ptr` approach lowers the overall transfer overhead here? Or is it due to ownership elsewhere?

File: pennylane_lightning/src/algorithms/StateVecAdjDiff.hpp
Comment (mlxd) [inline pennylane_lightning/src/algorithms/StateVecAdjDiff.hpp:55]: Nice work on this!

**Review by mlxd — COMMENTED** [review]
This looks amazing, great job @chaeyeunpark ! I have nothing else major to add, but a few comments.

The only thing I'd like to see before we get this merged in are some runtime plots comparing the before and after (v0.23.0, v0.24.0-dev, and this).

File: pennylane_lightning/src/algorithms/StateVecAdjDiff.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/StateVecAdjDiff.hpp:54]: Thx!

File: pennylane_lightning/src/util/Memory.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/util/Memory.hpp:71]: Yes. Moreover, as we use `clang-cl` on Windows (for Kokkos), we cannot use `pmr` even though MSVC supports it.

File: pennylane_lightning/src/algorithms/JacobianTape.hpp
Comment (chaeyeunpark) [reply pennylane_lightning/src/algorithms/JacobianTape.hpp:241]: When we serialize, Python directly creates an `std::shared_ptr<[a subclass of Observable<T>]>` object, e.g. `ob = NamedObsC64("X", [0]` creates `std::shared_ptr<NamedObs<double>>`. Thus we can just use it. It is possible (in principle) to use `unique_ptr` but a Python code should know that the object is consumed by the C++ code (which is rather complicated).

File: pennylane_lightning/src/util/Memory.hpp
Comment (mlxd) [reply pennylane_lightning/src/util/Memory.hpp:71]: Do they have it even in experimental support? We could wrap `std::experimental::pmr` if so.

File: pennylane_lightning/lightning_qubit.py
Comment (chaeyeunpark) [reply pennylane_lightning/lightning_qubit.py:297]: We just need a state after apply all operations (not measurement results such as `expval`, `var`, `samples`, etc.) This is what `adjoint_jacobian` also does: https://github.com/PennyLaneAI/pennylane/blob/b6fc5380abea6215661704ebe2f5cb8e7a599635/pennylane/_qubit_device.py#L1074-L1077

**Comment by chaeyeunpark** [comment]
```python
import pennylane as qml
import pennylane.numpy as np
from numpy.random import default_rng
import timeit
import sys

rng = default_rng(1337)
num_qubits = 10

def create_random_pauli_prod(num_qubits):
    paulis = [qml.Identity, qml.PauliX, qml.PauliY, qml.PauliZ]
    pindices = rng.integers(0, 3, num_qubits)
    terms = [paulis[pindices[i]](i) for i in range(num_qubits)]
    return qml.operation.Tensor(*terms)

def create_hamiltonian(num_qubits, n_terms):
    ham_coeffs = rng.standard_normal(n_terms)
    ham_terms = [create_random_pauli_prod(num_qubits) for i in range(n_terms)]
    return qml.Hamiltonian(ham_coeffs, ham_terms)

dev = qml.device('lightning.qubit', wires=num_qubits)

def bench_func(n_terms):
    ham = create_hamiltonian(num_qubits, n_terms)
    with qml.tape.QuantumTape() as tape:
        qml.RX(0.23, wires=0)
        qml.CNOT(wires=[0,1])
        qml.RY(0.23, wires=1)
        qml.CNOT(wires=[1,2])
        qml.RZ(0.23, wires=2)
        qml.expval(ham)

    w = tape.get_parameters(tranable_only = True)

    def timing_this():
        weights = rng.random(3)
        weights = np.array(weights)
        w[:3] = weights
        tape.set_parameters(w, trainable_only=True)
        return dev.adjoint_jacobian(tape, np.ones(2**num_qubits, requires_grad=False, dtype=np.complex128))

    return timeit.timeit(timing_this, number=100)

for n_terms in [2**k for k in range(2, 9)]:
    print("{}\t{}".format(n_terms, bench_func(n_terms)))
```

Using this simple benchmark code, I got
```
master version
4	19.39245017792564
8	19.997620621928945
16	22.724174362956546
32	26.514014064101502
64	33.93342457001563
128	54.66594595997594
256	104.8523806289304

this PR
4	0.30219985102303326
8	0.20092515798751265
16	0.445317056030035
32	1.591180496965535
64	8.25559274898842
128	19.637774570961483
256	34.774507706053555
```

The first column is the number of terms in the Hamiltonian.

**Comment by chaeyeunpark** [comment]
I also get slightly faster results for a simple observables: 
```python
import pennylane as qml
import pennylane.numpy as np
from numpy.random import default_rng
import timeit
import sys

rng = default_rng(1337)

def bench_func(num_qubits):
    dev = qml.device('lightning.qubit', wires=num_qubits)
    with qml.tape.QuantumTape() as tape:
        qml.RX(0.23, wires=0)
        qml.CNOT(wires=[0,1])
        qml.RY(0.23, wires=1)
        qml.CNOT(wires=[1,2])
        qml.RZ(0.23, wires=2)
        qml.expval(qml.PauliX(0) @ qml.PauliY(1))

    def timing_this():
        weights = rng.random(3)
        weights = np.array(weights)
        tape.set_parameters(weights, trainable_only=True)
        return dev.adjoint_jacobian(tape, np.ones(2**num_qubits, requires_grad=False, dtype=np.complex128))

    return timeit.timeit(timing_this, number=100)

for num_qubits in [4, 6, 8, 10, 12, 14, 16, 18]:
    print("{}\t{}".format(num_qubits, bench_func(num_qubits)))
```

And the result is:
```
master:
4	0.4656622860347852
6	0.2100854809395969
8	0.05818713700864464
10	0.6123739270260558
12	0.2523528659949079
14	0.19010203890502453
16	1.332373121054843
18	4.672340785968117

this PR:
4	0.12154659803491086
6	0.06435875396709889
8	0.061190371052362025
10	0.05885270098224282
12	0.17519149102736264
14	0.20414239005185664
16	0.935425164992921
18	4.56479615601711
```
The first column is the number of qubits.

**Comment by mlxd** [comment]
> I also get slightly faster results for a simple observables:
> 
> ```python
> import pennylane as qml
> import pennylane.numpy as np
> from numpy.random import default_rng
> import timeit
> import sys
> 
> rng = default_rng(1337)
> 
> def bench_func(num_qubits):
>     dev = qml.device('lightning.qubit', wires=num_qubits)
>     with qml.tape.QuantumTape() as tape:
>         qml.RX(0.23, wires=0)
>         qml.CNOT(wires=[0,1])
>         qml.RY(0.23, wires=1)
>         qml.CNOT(wires=[1,2])
>         qml.RZ(0.23, wires=2)
>         qml.expval(qml.PauliX(0) @ qml.PauliY(1))
> 
>     def timing_this():
>         weights = rng.random(3)
>         weights = np.array(weights)
>         tape.set_parameters(weights, trainable_only=True)
>         return dev.adjoint_jacobian(tape, np.ones(2**num_qubits, requires_grad=False, dtype=np.complex128))
> 
>     return timeit.timeit(timing_this, number=100)
> 
> for num_qubits in [4, 6, 8, 10, 12, 14, 16, 18]:
>     print("{}\t{}".format(num_qubits, bench_func(num_qubits)))
> ```
> 
> And the result is:
> 
> ```
> master:
> 4	0.4656622860347852
> 6	0.2100854809395969
> 8	0.05818713700864464
> 10	0.6123739270260558
> 12	0.2523528659949079
> 14	0.19010203890502453
> 16	1.332373121054843
> 18	4.672340785968117
> 
> this PR:
> 4	0.12154659803491086
> 6	0.06435875396709889
> 8	0.061190371052362025
> 10	0.05885270098224282
> 12	0.17519149102736264
> 14	0.20414239005185664
> 16	0.935425164992921
> 18	4.56479615601711
> ```
> 
> The first column is the number of qubits.

Those numbers are amazing --- excellent work putting this together!

File: tests/test_adjoint_jacobian.py
Comment (chaeyeunpark) [inline tests/test_adjoint_jacobian.py:825]: This is the test case @AmintorDusko found, which is fixed in this PR.

File: pennylane_lightning/lightning_qubit.py
Comment (chaeyeunpark) [inline pennylane_lightning/lightning_qubit.py:395]: And these are relevant lines.

**Review by mlxd — APPROVED** [review]
Nothing more to add. Great work, and thanks @chaeyeunpark

File: tests/test_adjoint_jacobian.py
Comment (AmintorDusko) [inline tests/test_adjoint_jacobian.py]:
```suggestion
    dev_lig = qml.device("lightning.qubit", wires=qubits, c_dtype = c_dtype)
```

Comment (AmintorDusko) [reply tests/test_adjoint_jacobian.py]: This custom decomposition is not necessary anymore.

File: tests/test_adjoint_jacobian.py
Comment (AmintorDusko) [reply tests/test_adjoint_jacobian.py:825]: Thank you for the quick fix and test inclusion!

File: pennylane_lightning/lightning_qubit.py
Comment (AmintorDusko) [reply pennylane_lightning/lightning_qubit.py:395]: Nice!

**Review by AmintorDusko — APPROVED** [review]
Excellent work! You have my 100% OK after the small fix, I suggest.

**Comment by chaeyeunpark** [comment]
Thanks @AmintorDusko and @mlxd for the reviews. I just added one small test. I will merge this PR when Tidy is satisfied with the code.

**Pull request merged**
